src/Microsoft.Identity.Web.TokenAcquisition/ConfidentialClientApplicationBuilderExtension.cs

```diff
@@ -5,55 +5,78 @@
 using System.Collections.Generic;
 using System.Linq;
 using System.Security.Cryptography.X509Certificates;
+using System.Threading;
+using Azure.Identity;
+using Microsoft.Extensions.Logging;
 using Microsoft.Identity.Client;
 
 namespace Microsoft.Identity.Web
 {
     internal static class ConfidentialClientApplicationBuilderExtension
     {
-        public static ConfidentialClientApplicationBuilder WithClientCredentials(this ConfidentialClientApplicationBuilder builder, IEnumerable<CredentialDescription> clientCredentials)
+        public static ConfidentialClientApplicationBuilder WithClientCredentials(this ConfidentialClientApplicationBuilder builder, IEnumerable<CredentialDescription> clientCredentials, ILogger logger)
         {
             foreach (var credential in clientCredentials)
             {
-                // TODO, use the credential loader
-                if (credential.SourceType == CredentialSource.SignedAssertionFromManagedIdentity)
+                if (!credential.Skip)
                 {
-                    if (credential.CachedValue == null)
+                    // TODO, use the credential loader
+                    if (credential.SourceType == CredentialSource.SignedAssertionFromManagedIdentity)
                     {
-                        credential.CachedValue = new ManagedIdentityClientAssertion(credential.ManagedIdentityClientId);
+                        ManagedIdentityClientAssertion? managedIdentityClientAssertion = credential.CachedValue as ManagedIdentityClientAssertion;
+                        if (credential.CachedValue == null)
+                        {
+                            managedIdentityClientAssertion = new ManagedIdentityClientAssertion(credential.ManagedIdentityClientId);
+                            credential.CachedValue = managedIdentityClientAssertion;
+                        }
+                        try
+                        {
+                            // Given that managed identity can be not available locally, we need to try to get a
+                            // signed assertion, and if it fails, move to the next credentials
+                            managedIdentityClientAssertion!.GetSignedAssertion(CancellationToken.None).GetAwaiter().GetResult();
+                        }
+                        catch (CredentialUnavailableException)
+                        {
+                            credential.Skip = true;
+                            logger.LogInformation("Not using Managed identity for client credentials");
```

Comment:
Use high performance logging here. I will do it. but merge this PR first and create an issue for me please to follow up.